### PR TITLE
[code-completion] Fix assertion hit in getTypeOfReference via code completion

### PR DESCRIPTION
**Pull request by nathawes**
```
enum QuxEnum : Int {
  case Qux1 = 10
  case Qux2 = 20
}

func foo(x: QuxEnum) {
  _ = x == .Qux1 || x == ./* code complete here */Qux2
}
```
Code completion is re-typechecking the assignment in `func foo` above. On the first pass the `UnresolvedMemberExpr`s corresponding to the enum elements are transformed into `DotSyntaxCallExpr`s. When it later re-typechecks via `typeCheckUnresolvedExpr` with the assignment as the parent, the type checker ends up in `getTypeOfReference` where references to `EnumElementDecl`s are assumed to not occur.  It looks like `typeCheckUnresolvedExpr` calls an overload of solve() that bypasses the SanitizeExpr ASTWalker, which cleans up ASTs for the constraint-based type checker. Adding that along this path seems to resolve the issue.

Resolves Resolves rdar://problem/38144409.

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please smoke test

**Review by xedin — COMMENTED** [review]
LGTM, Thanks!

**Review by xedin — APPROVED** [review]

**Review by benlangmuir — APPROVED** [review]
LGTM.  It's too bad you have to reconstruct the ast node but I guess we have no way to access the original one at this point.

**Comment by nathawes** [comment]
@swift-ci please smoke test

**Comment by nathawes** [comment]
The previous fix caused a regression for `CompletionKind::PostfixExpr` caught by a local run of the stress tester. Added the test case and changed the fix to focus on unresolved members specifically (`InferUnresolvedMemberConstraintGenerator` is only used when completing unresolved members from what I can tell).

Does this look ok @benlangmuir ?

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -3431,6 +3431,17 @@ class InferUnresolvedMemberConstraintGenerator : public ConstraintGenerator {
     return Expr->getType();
   }
 
+  Type visitDeclRefExpr(DeclRefExpr *Expr) override {
```
Comment (xedin) [inline lib/Sema/CSGen.cpp]: I think we should review what `InferUnresolvedMemberConstraintGenerator` is supposed to do and maybe instead just expect the number of callbacks we have for constraint generation...

File: lib/Sema/CSGen.cpp [hunk 2]
```diff
@@ -3431,6 +3431,17 @@ class InferUnresolvedMemberConstraintGenerator : public ConstraintGenerator {
...
+    if (Target != Expr) {
+      if (Type Ty = Expr->getType())
+        return Ty;
+      return ConstraintGenerator::visitDeclRefExpr(Expr);
+    }
+    // Otherwise, create a type variable saying we know nothing about this expr.
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: Instead of having a subclass for this purpose it could be a flag for the constraint generator

**Comment by xedin** [comment]
@nathawes Yes, that looks great!

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
Failures seem unrelated
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci smoke test OSX

**Comment by nathawes** [comment]
@swift-ci smoke test OSX

**Comment by nathawes** [comment]
@swift-ci please smoke test OS X

**Pull request merged**